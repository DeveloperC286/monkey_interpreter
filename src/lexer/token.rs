#[derive(PartialEq, PartialOrd)]
pub struct Token {
    pub token_type: TokenType,
    pub literal: String,
}

#[derive(PartialEq, PartialOrd)]
#[allow(non_camel_case_types)]
pub enum TokenType {
    ILLEGAL,
    EOF,

    // Identifiers
    IDENTIFIER,
    INTEGER,

    // Operators
    ASSIGN,
    PLUS,

    // Delimiters
    COMMA,
    SEMI_COLON,

    // Brackets
    OPENING_ROUND_BRACKET,
    CLOSING_ROUND_BRACKET,

    OPENING_CURLY_BRACKET,
    CLOSING_CURLY_BRACKET,

    OPENING_SQUARE_BRACKET,
    CLOSING_SQUARE_BRACKET,

    OPENING_ANGLE_BRACKET,
    CLOSING_ANGLE_BRACKET,

    //Keywords
    FUNCTION,
    LET,
}
