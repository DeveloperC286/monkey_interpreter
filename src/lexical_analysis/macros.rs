macro_rules! parse_characters {
    ($iterator:expr, $current_character:expr, $valid_character:ident) => {
        let mut characters = vec![$current_character];

        loop {
            match $iterator.peek() {
                Some(character) => {
                    if $valid_character(*character) {
                        characters.push($iterator.next().unwrap());
                    } else {
                        break;
                    }
                }
                None => {
                    break;
                }
            }
        }

        return String::from_iter(characters.iter());
    };
}

macro_rules! check_next_character {
    ($iterator:expr, $expected_next_character:expr, $token_type:expr) => {
        match $iterator.peek() {
            Some($expected_next_character) => {
                $iterator.next();
                return $token_type;
            }
            _ => {}
        }
    };
}
