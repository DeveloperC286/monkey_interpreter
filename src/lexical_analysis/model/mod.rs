pub(crate) mod lexical_error;
pub(crate) mod token;
