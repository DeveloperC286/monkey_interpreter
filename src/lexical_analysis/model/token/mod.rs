#[derive(Debug, Clone, PartialEq)]
pub(crate) enum Token {
    // Identifiers
    Identifier { literal: String },

    //Values
    Integer { literal: i64 },
    True,
    False,
    String { literal: String },

    // Operators
    Assign,
    Plus,
    Minus,
    Divide,
    Multiply,
    GreaterThan,
    LesserThan,
    Not,
    Equals,
    NotEquals,

    // Delimiters
    Comma,
    SemiColon,

    // Brackets
    OpeningRoundBracket,
    ClosingRoundBracket,

    OpeningCurlyBracket,
    ClosingCurlyBracket,

    OpeningSquareBracket,
    ClosingSquareBracket,

    //Keywords
    Function,
    Let,
    If,
    Else,
    Return,
}
