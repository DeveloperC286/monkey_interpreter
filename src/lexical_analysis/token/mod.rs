use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
#[allow(non_camel_case_types)]
pub enum Token {
    ILLEGAL { literal: String },
    EOF,

    // Identifiers
    IDENTIFIER { literal: String },
    INTEGER { literal: String },

    // Operators
    ASSIGN,
    PLUS,
    MINUS,
    DIVIDE,
    MULTIPLY,
    GREATER_THAN,
    LESSER_THAN,
    NOT,
    EQUALS,
    NOT_EQUALS,

    // Delimiters
    COMMA,
    SEMI_COLON,

    // Brackets
    OPENING_ROUND_BRACKET,
    CLOSING_ROUND_BRACKET,

    OPENING_CURLY_BRACKET,
    CLOSING_CURLY_BRACKET,

    //Keywords
    FUNCTION,
    LET,
    TRUE,
    FALSE,
    IF,
    ELSE,
    RETURN,
}
