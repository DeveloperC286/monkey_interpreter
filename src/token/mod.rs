#[derive(Debug, PartialEq)]
pub struct Token {
    pub token_type: TokenType,
    pub literal: String,
}

#[derive(Debug, Copy, Clone, PartialEq)]
#[allow(non_camel_case_types)]
pub enum TokenType {
    ILLEGAL,
    EOF,

    // Identifiers
    IDENTIFIER,
    INTEGER,

    // Operators
    ASSIGN,
    PLUS,
    MINUS,
    DIVIDE,
    MULTIPLY,
    GREATER_THAN,
    LESSER_THAN,

    // Delimiters
    COMMA,
    SEMI_COLON,

    // Brackets
    OPENING_ROUND_BRACKET,
    CLOSING_ROUND_BRACKET,

    OPENING_CURLY_BRACKET,
    CLOSING_CURLY_BRACKET,

    //Keywords
    FUNCTION,
    LET,
}
