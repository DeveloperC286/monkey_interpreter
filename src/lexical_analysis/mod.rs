use std::iter::{FromIterator, Peekable};
use std::str::Chars;

use crate::lexical_analysis::model::lexical_error::LexicalError;
use crate::lexical_analysis::model::token::Token;
use crate::lexical_analysis::utilities::*;

#[macro_use]
mod macros;

pub(crate) mod model;
mod utilities;

pub(crate) struct LexicalAnalysis<'a> {
    source_code: Peekable<Chars<'a>>,
}

impl<'a> LexicalAnalysis<'a> {
    pub(crate) fn from(code: &str) -> Result<Vec<Token>, LexicalError> {
        let mut lexical_analysis = LexicalAnalysis {
            source_code: code.chars().peekable(),
        };

        lexical_analysis.get_tokens()
    }

    fn get_tokens(&mut self) -> Result<Vec<Token>, LexicalError> {
        let mut tokens = Vec::new();

        loop {
            let token = self.get_next_token()?;
            let end_of_file = token == Token::EndOfFile;
            tokens.push(token);

            if end_of_file {
                break;
            }
        }

        Ok(tokens)
    }

    fn get_next_token(&mut self) -> Result<Token, LexicalError> {
        match self.get_next_character() {
            Some(character) => {
                debug!("Matching the character '{}'.", character);
                match character {
                    '!' => {
                        check_next_character!(self.source_code, '=', Ok(Token::NotEquals));
                        Ok(Token::Not)
                    }
                    '-' => Ok(Token::Minus),
                    '/' => Ok(Token::Divide),
                    '*' => Ok(Token::Multiply),
                    '>' => Ok(Token::GreaterThan),
                    '<' => Ok(Token::LesserThan),
                    '=' => {
                        check_next_character!(self.source_code, '=', Ok(Token::Equals));
                        Ok(Token::Assign)
                    }
                    '+' => Ok(Token::Plus),
                    '(' => Ok(Token::OpeningRoundBracket),
                    ')' => Ok(Token::ClosingRoundBracket),
                    '{' => Ok(Token::OpeningCurlyBracket),
                    '}' => Ok(Token::ClosingCurlyBracket),
                    ',' => Ok(Token::Comma),
                    ';' => Ok(Token::SemiColon),
                    _ => {
                        if is_valid_identifier_character(character) {
                            debug!("Parsing word from characters.");
                            return Ok(get_keyword_token(&self.get_word(character)));
                        }

                        if is_digit(character) {
                            debug!("Parsing integer from characters.");
                            return Ok(Token::Integer {
                                literal: self.get_integer(character)?,
                            });
                        }

                        Err(LexicalError::IllegalCharacter(character))
                    }
                }
            }
            None => Ok(Token::EndOfFile),
        }
    }

    fn get_integer_string(&mut self, character: char) -> String {
        parse_characters!(self.source_code, character, is_digit);
    }

    fn get_integer(&mut self, character: char) -> Result<i64, LexicalError> {
        let integer_string = self.get_integer_string(character);

        match integer_string.parse() {
            Ok(integer) => Ok(integer),
            Err(_) => Err(LexicalError::NonNumericIntegerString(integer_string)),
        }
    }

    fn get_word(&mut self, character: char) -> String {
        parse_characters!(self.source_code, character, is_valid_identifier_character);
    }

    fn get_next_character(&mut self) -> Option<char> {
        trace!("Getting next character.");
        let mut next_character = self.source_code.next();

        if let Some(character) = next_character {
            match character {
                ' ' | '\t' | '\n' | '\r' => {
                    trace!("Consuming formatting character.");
                    next_character = self.get_next_character();
                }
                _ => {}
            }
        }

        next_character
    }
}

#[cfg(test)]
mod tests;
