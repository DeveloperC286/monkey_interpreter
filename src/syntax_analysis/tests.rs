use super::super::lexical_analysis::token::{Token, TokenType};
use super::*;

use insta::assert_json_snapshot;
use rstest::rstest;

#[rstest(
    tokens,
    snapshot_name,
    case(
        vec![
            Token{token_type: TokenType::TRUE, literal: "TRUE".to_string()},
            Token{token_type: TokenType::SEMI_COLON, literal: ";".to_string()},
            Token{token_type: TokenType::EOF, literal: "".to_string()},
        ],
        "test_syntax_analysis_for_boolean_expression_case1"
    ),
    case(
        vec![
            Token{token_type: TokenType::FALSE, literal: "false".to_string()},
            Token{token_type: TokenType::EOF, literal: "".to_string()},
        ],
        "test_syntax_analysis_for_boolean_expression_case2"
    ),
)]
fn test_syntax_analysis_for_boolean_expression(tokens: Vec<Token>, snapshot_name: &str) {
    //when
    let returned_abstract_syntax_tree = SyntaxAnalysis::get_abstract_syntax_tree(tokens);

    //then
    assert_json_snapshot!(snapshot_name, returned_abstract_syntax_tree);
}

#[rstest(
    tokens_1,
    snapshot_name_1,
    tokens_2,
    snapshot_name_2,
    case(
        vec![
            Token{token_type: TokenType::LET, literal: "let".to_string()},
            Token{token_type: TokenType::IDENTIFIER, literal: "x".to_string()},
            Token{token_type: TokenType::ASSIGN, literal: "=".to_string()},
            Token{token_type: TokenType::INTEGER, literal: "5".to_string()},
            Token{token_type: TokenType::SEMI_COLON, literal: ";".to_string()},
            Token{token_type: TokenType::EOF, literal: "".to_string()},
        ],
        "test_syntax_analysis_for_let_statements_case1",
        vec![
            Token{token_type: TokenType::LET, literal: "let".to_string()},
            Token{token_type: TokenType::IDENTIFIER, literal: "x".to_string()},
            Token{token_type: TokenType::ASSIGN, literal: "=".to_string()},
            Token{token_type: TokenType::INTEGER, literal: "5".to_string()},
            Token{token_type: TokenType::SEMI_COLON, literal: ";".to_string()},
            Token{token_type: TokenType::LET, literal: "let".to_string()},
            Token{token_type: TokenType::IDENTIFIER, literal: "z".to_string()},
            Token{token_type: TokenType::ASSIGN, literal: "=".to_string()},
            Token{token_type: TokenType::INTEGER, literal: "7".to_string()},
            Token{token_type: TokenType::PLUS, literal: "+".to_string()},
            Token{token_type: TokenType::INTEGER, literal: "10".to_string()},
            Token{token_type: TokenType::SEMI_COLON, literal: ";".to_string()},
            Token{token_type: TokenType::EOF, literal: "".to_string()},
        ],
        "test_syntax_analysis_for_let_statements_case2"
    ),
)]
fn test_syntax_analysis_for_successive_parsing(
    tokens_1: Vec<Token>,
    snapshot_name_1: &str,
    tokens_2: Vec<Token>,
    snapshot_name_2: &str,
) {
    //when
    let returned_abstract_syntax_tree_1 = SyntaxAnalysis::get_abstract_syntax_tree(tokens_1);
    let returned_abstract_syntax_tree_2 = SyntaxAnalysis::get_abstract_syntax_tree(tokens_2);

    //then
    assert_json_snapshot!(snapshot_name_1, returned_abstract_syntax_tree_1);
    assert_json_snapshot!(snapshot_name_2, returned_abstract_syntax_tree_2);
}

#[rstest(
    tokens,
    snapshot_name,
    case(
        vec![
            Token{token_type: TokenType::INTEGER, literal: "15".to_string()},
            Token{token_type: TokenType::PLUS, literal: "+".to_string()},
            Token{token_type: TokenType::INTEGER, literal: "5".to_string()},
            Token{token_type: TokenType::SEMI_COLON, literal: ";".to_string()},
            Token{token_type: TokenType::EOF, literal: "".to_string()},
        ],
        "test_syntax_analysis_for_infix_expression_case1"
    ),
    case(
        vec![
            Token{token_type: TokenType::INTEGER, literal: "6".to_string()},
            Token{token_type: TokenType::MINUS, literal: "-".to_string()},
            Token{token_type: TokenType::INTEGER, literal: "6".to_string()},
            Token{token_type: TokenType::SEMI_COLON, literal: ";".to_string()},
            Token{token_type: TokenType::EOF, literal: "".to_string()},
        ],
        "test_syntax_analysis_for_infix_expression_case2"
    ),
    case(
        vec![
            Token{token_type: TokenType::INTEGER, literal: "3".to_string()},
            Token{token_type: TokenType::MULTIPLY, literal: "*".to_string()},
            Token{token_type: TokenType::INTEGER, literal: "4".to_string()},
            Token{token_type: TokenType::SEMI_COLON, literal: ";".to_string()},
            Token{token_type: TokenType::EOF, literal: "".to_string()},
        ],
        "test_syntax_analysis_for_infix_expression_case3"
    ),
    case(
        vec![
            Token{token_type: TokenType::INTEGER, literal: "5".to_string()},
            Token{token_type: TokenType::DIVIDE, literal: "/".to_string()},
            Token{token_type: TokenType::INTEGER, literal: "10".to_string()},
            Token{token_type: TokenType::SEMI_COLON, literal: ";".to_string()},
            Token{token_type: TokenType::EOF, literal: "".to_string()},
        ],
        "test_syntax_analysis_for_infix_expression_case4"
    ),
    case(
        vec![
            Token{token_type: TokenType::MINUS, literal: "-".to_string()},
            Token{token_type: TokenType::INTEGER, literal: "5".to_string()},
            Token{token_type: TokenType::MULTIPLY, literal: "*".to_string()},
            Token{token_type: TokenType::INTEGER, literal: "10".to_string()},
            Token{token_type: TokenType::SEMI_COLON, literal: ";".to_string()},
            Token{token_type: TokenType::EOF, literal: "".to_string()},
        ],
        "test_syntax_analysis_for_infix_expression_case5"
    ),
    case(
        vec![
            Token{token_type: TokenType::INTEGER, literal: "5".to_string()},
            Token{token_type: TokenType::MINUS, literal: "-".to_string()},
            Token{token_type: TokenType::INTEGER, literal: "10".to_string()},
            Token{token_type: TokenType::MULTIPLY, literal: "*".to_string()},
            Token{token_type: TokenType::INTEGER, literal: "10".to_string()},
            Token{token_type: TokenType::SEMI_COLON, literal: ";".to_string()},
            Token{token_type: TokenType::EOF, literal: "".to_string()},
        ],
        "test_syntax_analysis_for_infix_expression_case6"
    ),
    case(
        vec![
            Token{token_type: TokenType::INTEGER, literal: "10".to_string()},
            Token{token_type: TokenType::DIVIDE, literal: "/".to_string()},
            Token{token_type: TokenType::INTEGER, literal: "5".to_string()},
            Token{token_type: TokenType::PLUS, literal: "+".to_string()},
            Token{token_type: TokenType::INTEGER, literal: "1".to_string()},
            Token{token_type: TokenType::SEMI_COLON, literal: ";".to_string()},
            Token{token_type: TokenType::EOF, literal: "".to_string()},
        ],
        "test_syntax_analysis_for_infix_expression_case7"
    ),
)]
fn test_syntax_analysis_for_infix_expression(tokens: Vec<Token>, snapshot_name: &str) {
    //when
    let returned_abstract_syntax_tree = SyntaxAnalysis::get_abstract_syntax_tree(tokens);

    //then
    assert_json_snapshot!(snapshot_name, returned_abstract_syntax_tree);
}

#[rstest(
    tokens,
    snapshot_name,
    case(
        vec![
            Token{token_type: TokenType::NOT, literal: "!".to_string()},
            Token{token_type: TokenType::EOF, literal: "".to_string()},
        ],
        "test_syntax_analysis_for_prefix_expression_syntax_errors_case1"
    ),
    case(
        vec![
            Token{token_type: TokenType::MINUS, literal: "-".to_string()},
            Token{token_type: TokenType::SEMI_COLON, literal: ";".to_string()},
            Token{token_type: TokenType::EOF, literal: "".to_string()},
        ],
        "test_syntax_analysis_for_prefix_expression_syntax_errors_case2"
    ),
)]
fn test_syntax_analysis_for_prefix_expression_syntax_errors(
    tokens: Vec<Token>,
    snapshot_name: &str,
) {
    //when
    let returned_abstract_syntax_tree = SyntaxAnalysis::get_abstract_syntax_tree(tokens);

    //then
    assert_json_snapshot!(snapshot_name, returned_abstract_syntax_tree);
}

#[rstest(
    tokens,
    snapshot_name,
    case(
        vec![
            Token{token_type: TokenType::NOT, literal: "!".to_string()},
            Token{token_type: TokenType::INTEGER, literal: "15".to_string()},
            Token{token_type: TokenType::EOF, literal: "".to_string()},
        ],
        "test_syntax_analysis_for_prefix_expression_case1"
    ),
    case(
        vec![
            Token{token_type: TokenType::MINUS, literal: "-".to_string()},
            Token{token_type: TokenType::INTEGER, literal: "3".to_string()},
            Token{token_type: TokenType::SEMI_COLON, literal: ";".to_string()},
            Token{token_type: TokenType::EOF, literal: "".to_string()},
        ],
        "test_syntax_analysis_for_prefix_expression_case2"
    ),
)]
fn test_syntax_analysis_for_prefix_expression(tokens: Vec<Token>, snapshot_name: &str) {
    //when
    let returned_abstract_syntax_tree = SyntaxAnalysis::get_abstract_syntax_tree(tokens);

    //then
    assert_json_snapshot!(snapshot_name, returned_abstract_syntax_tree);
}

#[rstest(
    tokens,
    snapshot_name,
    case(
        vec![
            Token{token_type: TokenType::IDENTIFIER, literal: "temp".to_string()},
            Token{token_type: TokenType::SEMI_COLON, literal: ";".to_string()},
            Token{token_type: TokenType::EOF, literal: "".to_string()},
        ],
        "test_syntax_analysis_for_identifier_expression_case1"
    ),
    case(
        vec![
            Token{token_type: TokenType::IDENTIFIER, literal: "varX".to_string()},
            Token{token_type: TokenType::EOF, literal: "".to_string()},
        ],
        "test_syntax_analysis_for_identifier_expression_case2"
    ),
)]
fn test_syntax_analysis_for_identifier_expression(tokens: Vec<Token>, snapshot_name: &str) {
    //when
    let returned_abstract_syntax_tree = SyntaxAnalysis::get_abstract_syntax_tree(tokens);

    //then
    assert_json_snapshot!(snapshot_name, returned_abstract_syntax_tree);
}

#[rstest(
    tokens,
    snapshot_name,
    case(
        vec![
            Token{token_type: TokenType::SEMI_COLON, literal: "%".to_string()},
            Token{token_type: TokenType::EOF, literal: "".to_string()},
        ],
        "test_syntax_analysis_for_expression_syntax_errors_case1"
    ),
)]
fn test_syntax_analysis_for_expression_syntax_errors(tokens: Vec<Token>, snapshot_name: &str) {
    //when
    let returned_abstract_syntax_tree = SyntaxAnalysis::get_abstract_syntax_tree(tokens);

    //then
    assert_json_snapshot!(snapshot_name, returned_abstract_syntax_tree);
}

#[rstest(
    tokens,
    snapshot_name,
    case(
        vec![
            Token{token_type: TokenType::INTEGER, literal: "5".to_string()},
            Token{token_type: TokenType::SEMI_COLON, literal: ";".to_string()},
            Token{token_type: TokenType::EOF, literal: "".to_string()},
        ],
        "test_syntax_analysis_for_integer_expression_case1"
    ),
    case(
        vec![
            Token{token_type: TokenType::INTEGER, literal: "12".to_string()},
            Token{token_type: TokenType::EOF, literal: "".to_string()},
        ],
        "test_syntax_analysis_for_integer_expression_case2"
    ),
)]
fn test_syntax_analysis_for_integer_expression(tokens: Vec<Token>, snapshot_name: &str) {
    //when
    let returned_abstract_syntax_tree = SyntaxAnalysis::get_abstract_syntax_tree(tokens);

    //then
    assert_json_snapshot!(snapshot_name, returned_abstract_syntax_tree);
}

#[rstest(
    tokens,
    snapshot_name,
    case(
        vec![
            Token{token_type: TokenType::LET, literal: "let".to_string()},
            Token{token_type: TokenType::IDENTIFIER, literal: "x".to_string()},
            Token{token_type: TokenType::ASSIGN, literal: "=".to_string()},
            Token{token_type: TokenType::INTEGER, literal: "5".to_string()},
            Token{token_type: TokenType::SEMI_COLON, literal: ";".to_string()},
            Token{token_type: TokenType::RETURN, literal: "RETURN".to_string()},
            Token{token_type: TokenType::INTEGER, literal: "5".to_string()},
            Token{token_type: TokenType::SEMI_COLON, literal: ";".to_string()},
            Token{token_type: TokenType::EOF, literal: "".to_string()},
        ],
        "test_syntax_analysis_for_combined_statements_case1"
    ),
)]
fn test_syntax_analysis_for_combined_statements(tokens: Vec<Token>, snapshot_name: &str) {
    //when
    let returned_abstract_syntax_tree = SyntaxAnalysis::get_abstract_syntax_tree(tokens);

    //then
    assert_json_snapshot!(snapshot_name, returned_abstract_syntax_tree);
}

#[rstest(
    tokens,
    snapshot_name,
    case(
        vec![
            Token{token_type: TokenType::RETURN, literal: "RETURN".to_string()},
            Token{token_type: TokenType::INTEGER, literal: "5".to_string()},
            Token{token_type: TokenType::SEMI_COLON, literal: ";".to_string()},
            Token{token_type: TokenType::EOF, literal: "".to_string()},
        ],
        "test_syntax_analysis_for_return_statements_case1"
    ),
    case(
         vec![
            Token{token_type: TokenType::RETURN, literal: "RETURN".to_string()},
            Token{token_type: TokenType::INTEGER, literal: "5".to_string()},
            Token{token_type: TokenType::SEMI_COLON, literal: ";".to_string()},
            Token{token_type: TokenType::RETURN, literal: "return".to_string()},
            Token{token_type: TokenType::IDENTIFIER, literal: "x".to_string()},
            Token{token_type: TokenType::SEMI_COLON, literal: ";".to_string()},
            Token{token_type: TokenType::EOF, literal: "".to_string()},
        ],
        "test_syntax_analysis_for_return_statements_case2"
    ),
)]
fn test_syntax_analysis_for_return_statements(tokens: Vec<Token>, snapshot_name: &str) {
    //when
    let returned_abstract_syntax_tree = SyntaxAnalysis::get_abstract_syntax_tree(tokens);

    //then
    assert_json_snapshot!(snapshot_name, returned_abstract_syntax_tree);
}

#[rstest(
    tokens,
    snapshot_name,
    case(
        vec![
            Token{token_type: TokenType::LET, literal: "let".to_string()},
            Token{token_type: TokenType::IDENTIFIER, literal: "x".to_string()},
            Token{token_type: TokenType::ASSIGN, literal: "=".to_string()},
            Token{token_type: TokenType::INTEGER, literal: "5".to_string()},
            Token{token_type: TokenType::SEMI_COLON, literal: ";".to_string()},
            Token{token_type: TokenType::EOF, literal: "".to_string()},
        ],
        "test_syntax_analysis_for_let_statements_case1"
    ),
    case(
        vec![
            Token{token_type: TokenType::LET, literal: "let".to_string()},
            Token{token_type: TokenType::IDENTIFIER, literal: "x".to_string()},
            Token{token_type: TokenType::ASSIGN, literal: "=".to_string()},
            Token{token_type: TokenType::INTEGER, literal: "5".to_string()},
            Token{token_type: TokenType::SEMI_COLON, literal: ";".to_string()},
            Token{token_type: TokenType::LET, literal: "let".to_string()},
            Token{token_type: TokenType::IDENTIFIER, literal: "z".to_string()},
            Token{token_type: TokenType::ASSIGN, literal: "=".to_string()},
            Token{token_type: TokenType::INTEGER, literal: "7".to_string()},
            Token{token_type: TokenType::PLUS, literal: "+".to_string()},
            Token{token_type: TokenType::INTEGER, literal: "10".to_string()},
            Token{token_type: TokenType::SEMI_COLON, literal: ";".to_string()},
            Token{token_type: TokenType::EOF, literal: "".to_string()},
        ],
        "test_syntax_analysis_for_let_statements_case2"
    ),
)]
fn test_syntax_analysis_for_let_statements(tokens: Vec<Token>, snapshot_name: &str) {
    //when
    let returned_abstract_syntax_tree = SyntaxAnalysis::get_abstract_syntax_tree(tokens);

    //then
    assert_json_snapshot!(snapshot_name, returned_abstract_syntax_tree);
}

#[rstest(
    tokens,
    snapshot_name,
    case(
        vec![
            Token{token_type: TokenType::LET, literal: "let".to_string()},
            Token{token_type: TokenType::IDENTIFIER, literal: "x".to_string()},
            Token{token_type: TokenType::EOF, literal: "".to_string()},
        ],
        "test_syntax_analysis_for_let_statements_syntax_errors_case1"
    ),
    case(
        vec![
            Token{token_type: TokenType::LET, literal: "let".to_string()},
            Token{token_type: TokenType::EOF, literal: "".to_string()},
        ],
        "test_syntax_analysis_for_let_statements_syntax_errors_case2"
    ),
)]
fn test_syntax_analysis_for_let_statements_syntax_errors(tokens: Vec<Token>, snapshot_name: &str) {
    //when
    let returned_abstract_syntax_tree = SyntaxAnalysis::get_abstract_syntax_tree(tokens);

    //then
    assert_json_snapshot!(snapshot_name, returned_abstract_syntax_tree);
}

#[rstest(
    tokens,
    snapshot_name,
    case(
        vec![
            Token{token_type: TokenType::EOF, literal: "".to_string()},
        ],
        "test_syntax_analysis_tokens_input_edgecases_case1"
    ),
    case(
        vec![],
        "test_syntax_analysis_tokens_input_edgecases_case2"
    ),
)]
fn test_syntax_analysis_tokens_input_edgecases(tokens: Vec<Token>, snapshot_name: &str) {
    //when
    let returned_abstract_syntax_tree = SyntaxAnalysis::get_abstract_syntax_tree(tokens);

    //then
    assert_json_snapshot!(snapshot_name, returned_abstract_syntax_tree);
}
