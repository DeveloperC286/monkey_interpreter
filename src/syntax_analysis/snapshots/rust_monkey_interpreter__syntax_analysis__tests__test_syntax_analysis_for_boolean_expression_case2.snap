---
source: src/syntax_analysis/tests.rs
expression: returned_abstract_syntax_tree
---
{
  "program": [
    {
      "EXPRESSION_STATEMENT": {
        "initial_expression_token": {
          "token_type": "FALSE",
          "literal": "false"
        },
        "expression": {
          "BOOLEAN": {
            "boolean_token": {
              "token_type": "FALSE",
              "literal": "false"
            }
          }
        }
      }
    }
  ],
  "syntax_parsing_errors": []
}
